File: Backend/backend/services/email_service.go
package services

import (
	"crypto/rand"
	"fmt"
	"log"
	"math/big"
	"net/smtp"
	"strings"
)

// EmailService sends OTP and notification emails via SMTP.
//
// If SMTP credentials are not configured, falls back to logging the email
// content to stdout — useful for local development without a Gmail account.
type EmailService struct {
	host     string
	port     string
	username string
	password string
	from     string
}

func NewEmailService(host, port, username, password, from string) *EmailService {
	if from == "" {
		from = username
	}
	return &EmailService{
		host:     host,
		port:     port,
		username: username,
		password: password,
		from:     from,
	}
}

// GenerateOTP returns a 6-digit numeric one-time password.
func GenerateOTP() string {
	var sb strings.Builder
	for i := 0; i < 6; i++ {
		n, err := rand.Int(rand.Reader, big.NewInt(10))
		if err != nil {
			// extremely unlikely; pad with 0
			sb.WriteByte('0')
			continue
		}
		sb.WriteString(n.String())
	}
	return sb.String()
}

// SendOTP delivers a 6-digit OTP to `to`. Returns nil on success.
// If SMTP is not configured, prints to log and returns nil (dev fallback).
func (e *EmailService) SendOTP(to, code, lang string) error {
	subject, body := otpEmailContent(code, lang)
	return e.send(to, subject, body)
}

func (e *EmailService) send(to, subject, body string) error {
	if e.username == "" || e.password == "" {
		log.Printf("[email-dev] To: %s | Subject: %s\n%s", to, subject, body)
		return nil
	}
	addr := fmt.Sprintf("%s:%s", e.host, e.port)
	auth := smtp.PlainAuth("", e.username, e.password, e.host)
	msg := []byte(fmt.Sprintf(
		"From: %s\r\nTo: %s\r\nSubject: %s\r\nMIME-Version: 1.0\r\nContent-Type: text/plain; charset=\"utf-8\"\r\n\r\n%s",
		e.from, to, subject, body,
	))
	if err := smtp.SendMail(addr, auth, e.username, []string{to}, msg); err != nil {
		return fmt.Errorf("smtp send: %w", err)
	}
	return nil
}

func otpEmailContent(code, lang string) (subject, body string) {
	if lang == "ru" {
		return "Moshn — Код подтверждения",
			fmt.Sprintf("Здравствуйте!\n\nВаш код подтверждения: %s\n\nКод действителен 10 минут.\n\nЕсли вы не запрашивали этот код, проигнорируйте письмо.\n\n— Moshn", code)
	}
	return "Moshn — Tasdiqlash kodi",
		fmt.Sprintf("Salom!\n\nSizning tasdiqlash kodingiz: %s\n\nKod 10 daqiqa amal qiladi.\n\nAgar siz so'ramagan bo'lsangiz, ushbu xatni e'tiborsiz qoldiring.\n\n— Moshn", code)
}

File: Backend/nest-backend/src/email/email.service.ts
import { Injectable, Logger } from '@nestjs/common';
import * as nodemailer from 'nodemailer';

@Injectable()
export class EmailService {
  private readonly logger = new Logger(EmailService.name);
  private transporter: nodemailer.Transporter | null = null;

  constructor() {
    const user = process.env.SMTP_USER;
    const pass = process.env.SMTP_PASSWORD;

    if (user && pass) {
      this.transporter = nodemailer.createTransport({
        host: process.env.SMTP_HOST ?? 'smtp.gmail.com',
        port: Number(process.env.SMTP_PORT ?? 587),
        secure: false,
        auth: { user, pass },
      });
    }
  }

  async sendOtp(to: string, code: string, lang: string): Promise<void> {
    const [subject, body] = this.otpContent(code, lang);
    await this.send(to, subject, body);
  }

  private async send(to: string, subject: string, body: string): Promise<void> {
    if (!this.transporter) {
      this.logger.log(`[email-dev] To: ${to} | Subject: ${subject}\n${body}`);
      return;
    }
    const from = process.env.SMTP_FROM ?? process.env.SMTP_USER;
    await this.transporter.sendMail({ from, to, subject, text: body });
  }

  private otpContent(code: string, lang: string): [string, string] {
    if (lang === 'ru') {
      return [
        'Shina24 — Код подтверждения',
        `Здравствуйте!\n\nВаш код подтверждения: ${code}\n\nКод действителен 10 минут.\n\n— Shina24`,
      ];
    }
    return [
      'Shina24 — Tasdiqlash kodi',
      `Salom!\n\nSizning tasdiqlash kodingiz: ${code}\n\nKod 10 daqiqa amal qiladi.\n\n— Shina24`,
    ];
  }

  static generateOtp(): string {
    let code = '';
    for (let i = 0; i < 6; i++) {
      code += Math.floor(Math.random() * 10).toString();
    }
    return code;
  }
}
